### ISSUE TITLE
facelet loss3 fluctuated and trainging failed

### ISSUE THREAD

**Issue by pangwong** [yingcong/Facelet_Bank#13]
Hi yingcong, 
I am trying to reproduce your result  but got a waired problem, that the loss3 varied in big fluctuation when training. On the begining, everying seemed OK but the loss were too small. After training of several epoches, loss1 and loss2 stayed small, but loss3 began to fluctuate.  The sum loss is dominated by loss3 thus.
```shell
saving facehair_epoch_10.pth in /data/image_server/extra/tasks/gan_expr/facelet_bank/train/celeba_0819/snapshot_0819_sub/facehair_epoch_10.pth
  1%|▊                                                                                                                        | 5/695 [00:06<16:04,  1.40s/it](epoch: 11, iters: 5) loss1: 0.059 loss2: 0.060 loss3: 27.433 loss_all: 27.551
  2%|██▌                                                                                                                     | 15/695 [00:19<14:33,  1.28s/it](epoch: 11, iters: 15) loss1: 0.058 loss2: 0.057 loss3: 6.530 loss_all: 6.645
  4%|████▎                                                                                                                   | 25/695 [00:31<14:08,  1.27s/it](epoch: 11, iters: 25) loss1: 0.059 loss2: 0.060 loss3: 5.223 loss_all: 5.341
  5%|██████                                                                                                                  | 35/695 [00:44<13:50,  1.26s/it](epoch: 11, iters: 35) loss1: 0.058 loss2: 0.058 loss3: 11.128 loss_all: 11.245
  6%|███████▊                                                                                                                | 45/695 [00:56<13:35,  1.25s/it](epoch: 11, iters: 45) loss1: 0.059 loss2: 0.057 loss3: 8.589 loss_all: 8.706
  8%|█████████▍                                                                                                              | 55/695 [01:08<13:21,  1.25s/it](epoch: 11, iters: 55) loss1: 0.057 loss2: 0.056 loss3: 6.787 loss_all: 6.899
  9%|███████████▏                                                                                                            | 65/695 [01:21<13:08,  1.25s/it](epoch: 11, iters: 65) loss1: 0.057 loss2: 0.060 loss3: 18.024 loss_all: 18.142
 11%|████████████▉                                                                                                           | 75/695 [01:33<12:55,  1.25s/it](epoch: 11, iters: 75) loss1: 0.056 loss2: 0.056 loss3: 8.630 loss_all: 8.742
 12%|██████████████▋                                                                                                         | 85/695 [01:46<12:42,  1.25s/it](epoch: 11, iters: 85) loss1: 0.059 loss2: 0.061 loss3: 5.404 loss_all: 5.524
 14%|████████████████▍                                                                                                       | 95/695 [01:58<12:28,  1.25s/it](epoch: 11, iters: 95) loss1: 0.060 loss2: 0.061 loss3: 4.773 loss_all: 4.894
 15%|█████████████████▉                                                                                                     | 105/695 [02:10<12:15,  1.25s/it](epoch: 11, iters: 105) loss1: 0.190 loss2: 0.290 loss3: 960.732 loss_all: 961.213
 17%|███████████████████▋                                                                                                   | 115/695 [02:23<12:02,  1.25s/it](epoch: 11, iters: 115) loss1: 2.782 loss2: 10.159 loss3: 7643.598 loss_all: 7656.540
 18%|█████████████████████▍                                                                                                 | 125/695 [02:35<11:50,  1.25s/it](epoch: 11, iters: 125) loss1: 5.948 loss2: 24.127 loss3: 5259.168 loss_all: 5289.244
 19%|███████████████████████                                                                                                | 135/695 [02:48<11:37,  1.25s/it](epoch: 11, iters: 135) loss1: 3.336 loss2: 11.013 loss3: 3519.050 loss_all: 3533.399
 21%|████████████████████████▊                                                                                              | 145/695 [03:00<11:24,  1.25s/it](epoch: 11, iters: 145) loss1: 2.472 loss2: 11.408 loss3: 1584.675 loss_all: 1598.554
 22%|██████████████████████████▌                                                                                            | 155/695 [03:12<11:12,  1.25s/it](epoch: 11, iters: 155) loss1: 0.942 loss2: 2.853 loss3: 661.282 loss_all: 665.077
saving facehair_epoch_11.pth in /data/image_server/extra/tasks/gan_expr/facelet_bank/train/celeba_0819/snapshot_0819_sub/facehair_epoch_11.pth
 24%|████████████████████████████▎                                                                                          | 165/695 [03:26<11:02,  1.25s/it](epoch: 11, iters: 165) loss1: 1.397 loss2: 3.483 loss3: 487.248 loss_all: 492.128
 25%|█████████████████████████████▉                                                                                         | 175/695 [03:38<10:49,  1.25s/it](epoch: 11, iters: 175) loss1: 0.617 loss2: 2.370 loss3: 211.403 loss_all: 214.389
 27%|███████████████████████████████▋                                                                                       | 185/695 [03:50<10:36,  1.25s/it](epoch: 11, iters: 185) loss1: 0.289 loss2: 1.072 loss3: 201.949 loss_all: 203.310
 28%|█████████████████████████████████▍                                                                                     | 195/695 [04:03<10:23,  1.25s/it](epoch: 11, iters: 195) loss1: 0.326 loss2: 0.900 loss3: 74.685 loss_all: 75.911
 29%|███████████████████████████████████                                                                                    | 205/695 [04:15<10:11,  1.25s/it](epoch: 11, iters: 205) loss1: 0.514 loss2: 1.453 loss3: 183.029 loss_all: 184.997
 31%|████████████████████████████████████▊                                                                                  | 215/695 [04:28<09:58,  1.25s/it](epoch: 11, iters: 215) loss1: 0.288 loss2: 0.768 loss3: 183.889 loss_all: 184.946
 32%|██████████████████████████████████████▌                                                                                | 225/695 [04:40<09:45,  1.25s/it](epoch: 11, iters: 225) loss1: 0.307 loss2: 0.902 loss3: 66.429 loss_all: 67.638
 34%|████████████████████████████████████████▏                                                                              | 235/695 [04:52<09:33,  1.25s/it](epoch: 11, iters: 235) loss1: 0.291 loss2: 1.023 loss3: 135.353 loss_all: 136.666
 35%|█████████████████████████████████████████▉                                                                             | 245/695 [05:05<09:20,  1.25s/it](epoch: 11, iters: 245) loss1: 0.234 loss2: 0.553 loss3: 56.823 loss_all: 57.609
 37%|███████████████████████████████████████████▋                                                                           | 255/695 [05:17<09:08,  1.25s/it](epoch: 11, iters: 255) loss1: 0.157 loss2: 0.277 loss3: 152.729 loss_all: 153.163
 38%|█████████████████████████████████████████████▎                                                                         | 265/695 [05:30<08:55,  1.25s/it](epoch: 11, iters: 265) loss1: 0.192 loss2: 0.543 loss3: 19.501 loss_all: 20.236
 ```
(drag the bar right and you will see the loss value)

Although the training process could go to the end, the final result is not as expected. I train the model to add facehair using the ```sub vector``` 【dst_vector - src_vector】 as label. Other hyperparameters are default. The bad case is as follows:

![greyson_sample](https://user-images.githubusercontent.com/7804678/44378163-a0d40f00-a532-11e8-8b8c-26dcac6d22be.png)
![greyson_sample-facehair-s-5 0-1534815447](https://user-images.githubusercontent.com/7804678/44378167-a3ceff80-a532-11e8-8ad3-8182192891d9.png)

Note that, when I tried to train using the ```normialized vector```  【sub_vector / (sub_vector)**2.mean()】, I got acceptable result show below, not as good as your result though.  

![greyson_sample-facehair-s-0 5-1534815365](https://user-images.githubusercontent.com/7804678/44378320-6323b600-a533-11e8-9b16-273942adb920.png)

**I wonder whether you have met similar problem when you implemented your paper.**
Thank for your reply.

**Comment by pangwong** [comment]
I have tried to decrease the learning rate to 0.00005, but it doesn't get better.

**Comment by yingcong** [comment]
It seems strange. I haven't encountered this problem before. The training is simply trying to regress a pseudo label with a 3-layers network, the optimization should not be so difficult.

Perhaps you could debug the network with in the following ways:
1. check if the pseudo label is correct by visualizing its norm across channels, i.e., torch.norm(target, dim=1)

2. if you have used batch norm or instance norm for the facelet network, try to remove it.

**Comment by pangwong** [comment]
Thanks for your reply. I finally find that there is problem with my pesudo label. My bad.

**Comment by AidasK** [comment]
Could you post what results you achieved? Are they less blurry?

**Comment by AidasK** [comment]
![kid-goatee-s-0 45](https://user-images.githubusercontent.com/2088484/44575150-0246e880-a794-11e8-8c5e-5c2c0050ab00.png)
This is my blurry result with normalised vector

**Comment by pangwong** [comment]
@AidasK I did't get the same result with yingcong's model after training using about 3000 unaligned images and pesudo labels. below is my results with strength 5:
![0824_sub_greyson_sample-facehair-s-5 0-1535074370](https://user-images.githubusercontent.com/7804678/44577152-f3c3f580-a7c2-11e8-9025-3195ccf4570c.png)

I an going to test whether it will go better when training with aligned images as yingcong said in readme.

would you mind telling us how many image you used to train the provided facehair.pth model @yingcong

**Comment by AidasK** [comment]
I guess you got this result with aligned images, original facehair model generates same blurry results.
![kid-facehair-s-5 00](https://user-images.githubusercontent.com/2088484/44580741-c0be3980-a7a3-11e8-994f-4b175461f963.png)
I think this blur comes from the decoder. I can provide you with a script for decoder training if you are interested in spending more time on this and publishing your results.

**Comment by pangwong** [comment]
I got the result with Not Aligned images. As for the blurry results, maybe @yingcong  can give us an explanation.
